analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/ConditionEvaluatesToConstantBase.cs

```diff
@@ -63,21 +75,103 @@ public override void ExecutionCompleted()
+    {
+        var issueMessage = operation.Kind == OperationKindEx.IsNull ? MessageNullCoalescing : string.Format(MessageBool, conditionIsTrue);
+        var secondaryLocations = SecondaryLocations(block, conditionIsTrue);
+        if (secondaryLocations.Any())
+        {
+            ReportIssue(Rule2583, operation, secondaryLocations, string.Format(MessageUnreachable, issueMessage));
+        }
+        else
+        {
+            ReportIssue(Rule2589, operation, null, string.Format(issueMessage, conditionIsTrue));
+        }
+    }
+
+    private List<Location> SecondaryLocations(BasicBlock block, bool conditionIsTrue)
+    {
+        List<Location> locations = new();
+        IOperation currentStart = null;
+        IOperation currentEnd = null;
+        var unreachable = UnreachableOperations(block, conditionIsTrue).Where(x => !IsIgnoredLocation(x.Syntax));
+        foreach (var operation in unreachable.Concat(reached).OrderBy(x => x.Syntax.SpanStart))
+        {
+            if (unreachable.Contains(operation))
+            {
+                currentStart ??= operation;
+                currentEnd = operation;
+            }
+            else
+            {
+                if (currentStart is not null)
+                {
+                    locations.Add(currentStart.Syntax.CreateLocation(currentEnd.Syntax));
+                    currentStart = null;
+                }
+            }
+        }
+        if (currentStart != null)
+        {
+            locations.Add(currentStart.Syntax.CreateLocation(currentEnd.Syntax));
+        }
+        return locations;
+    }
+
+    private IEnumerable<IOperation> UnreachableOperations(BasicBlock block, bool conditionIsTrue)
     {
-        var issueMessage = operation.Kind == OperationKindEx.IsNull ? MessageNullCoalescing : string.Format(MessageBool, conditionEvaluation);
-        ReportIssue(Rule2589, operation, null, issueMessage);
+        if (block.SuccessorBlocks.Distinct().Count() != 2)
+        {
+            return Enumerable.Empty<IOperation>();
+        }
+        HashSet<BasicBlock> reachable = new() { block };
+        HashSet<BasicBlock> unreachable = new();
+
+        var reachableSuccessor = conditionIsTrue ^ block.ConditionKind == ControlFlowConditionKind.WhenFalse
+            ? block.ConditionalSuccessor.Destination
+            : block.SuccessorBlocks.Single(x => x != block.ConditionalSuccessor.Destination);
+        TraverseReachable(reachableSuccessor);
+        var unreachableSuccessor = block.SuccessorBlocks.Single(x => x != reachableSuccessor);
+        TraverseUnreachable(unreachableSuccessor);
+        return unreachable.SelectMany(x => x.OperationsAndBranchValue).Except(reached);
+
+        void TraverseReachable(BasicBlock block)
+        {
+            if (reachable.Add(block))
+            {
+                foreach (var successor in block.SuccessorBlocks)
+                {
+                    TraverseReachable(successor);
```

Comment:
I would avoid recursion here. We've seen user code that had thousands of autogenerated `flag == Flag.A || flag == flag.B`... and we'd run out of stack there. Please rewrite it using a queue (that's also what Roslyn typically does in these graph algorithms)